diff --git a/api/src/main/java/com/bytenya/minecraft/velocityguard/api/event/VgLoginEvent.java b/api/src/main/java/com/bytenya/minecraft/velocityguard/api/event/VgLoginEvent.java
new file mode 100644
index 00000000..bd4b993b
--- /dev/null
+++ b/api/src/main/java/com/bytenya/minecraft/velocityguard/api/event/VgLoginEvent.java
@@ -0,0 +1,38 @@
+package com.bytenya.minecraft.velocityguard.api.event;
+
+import com.google.common.base.Preconditions;
+import com.velocitypowered.api.event.ResultedEvent;
+import com.velocitypowered.api.event.annotation.AwaitingEvent;
+import com.velocitypowered.api.proxy.Player;
+
+@AwaitingEvent
+public class VgLoginEvent {
+
+  private final Player player;
+  private ResultedEvent.ComponentResult result;
+
+  public VgLoginEvent(Player player) {
+    this.player = Preconditions.checkNotNull(player, "player");
+    this.result = ResultedEvent.ComponentResult.allowed();
+  }
+
+  public Player getPlayer() {
+    return player;
+  }
+
+  public ResultedEvent.ComponentResult getResult() {
+    return result;
+  }
+
+  public void setResult(ResultedEvent.ComponentResult result) {
+    this.result = Preconditions.checkNotNull(result, "result");
+  }
+
+  @Override
+  public String toString() {
+    return "VgLoginEvent{"
+        + "player=" + player
+        + ", result=" + result
+        + '}';
+  }
+}
