diff --git a/api/src/main/java/com/bytenya/minecraft/velocityguard/api/event/VgPostLoginEvent.java b/api/src/main/java/com/bytenya/minecraft/velocityguard/api/event/VgPostLoginEvent.java
new file mode 100644
index 00000000..a7a62798
--- /dev/null
+++ b/api/src/main/java/com/bytenya/minecraft/velocityguard/api/event/VgPostLoginEvent.java
@@ -0,0 +1,25 @@
+package com.bytenya.minecraft.velocityguard.api.event;
+
+import com.google.common.base.Preconditions;
+import com.velocitypowered.api.event.annotation.AwaitingEvent;
+import com.velocitypowered.api.proxy.Player;
+
+@AwaitingEvent
+public class VgPostLoginEvent {
+  private final Player player;
+
+  public VgPostLoginEvent(Player player) {
+    this.player = Preconditions.checkNotNull(player, "player");
+  }
+
+  public Player getPlayer() {
+    return player;
+  }
+
+  @Override
+  public String toString() {
+    return "VgPostLoginEvent{"
+        + "player=" + player
+        + '}';
+  }
+}
