diff --git a/proxy/src/main/java/com/velocitypowered/proxy/connection/backend/BackendPlaySessionHandler.java b/proxy/src/main/java/com/velocitypowered/proxy/connection/backend/BackendPlaySessionHandler.java
index 128d5c37..b653dc3b 100644
--- a/proxy/src/main/java/com/velocitypowered/proxy/connection/backend/BackendPlaySessionHandler.java
+++ b/proxy/src/main/java/com/velocitypowered/proxy/connection/backend/BackendPlaySessionHandler.java
@@ -294,10 +294,12 @@ public class BackendPlaySessionHandler implements MinecraftSessionHandler {
     }
 
     if (PluginMessageUtil.isMcBrand(packet)) {
-      PluginMessagePacket rewritten = PluginMessageUtil
-              .rewriteMinecraftBrand(packet,
-                      server.getVersion(), playerConnection.getProtocolVersion());
-      playerConnection.write(rewritten);
+      // ByteNya start
+      // PluginMessagePacket rewritten = PluginMessageUtil
+      //         .rewriteMinecraftBrand(packet,
+      //                 server.getVersion(), playerConnection.getProtocolVersion());
+      // playerConnection.write(rewritten);
+      // ByteNya end
       return true;
     }
 
@@ -446,6 +448,13 @@ public class BackendPlaySessionHandler implements MinecraftSessionHandler {
     if (packet instanceof PluginMessagePacket) {
       ((PluginMessagePacket) packet).retain();
     }
+    // ByteNya start
+    if (packet instanceof UpsertPlayerInfoPacket) {
+      packet = server.vg.eventFactory.processPlayerInfoPacket(playerConnection, (UpsertPlayerInfoPacket) packet);
+    } else if (packet instanceof RemovePlayerInfoPacket) {
+      packet = server.vg.eventFactory.processPlayerInfoPacket(playerConnection, (RemovePlayerInfoPacket) packet);
+    }
+    // ByteNya end
     playerConnection.delayedWrite(packet);
     if (++packetsFlushed >= MAXIMUM_PACKETS_TO_FLUSH) {
       playerConnection.flush();
