diff --git a/proxy/src/main/java/com/bytenya/minecraft/velocityguard/limbo/server/commands/CmdConn.java b/proxy/src/main/java/com/bytenya/minecraft/velocityguard/limbo/server/commands/CmdConn.java
new file mode 100644
index 00000000..cdba83a7
--- /dev/null
+++ b/proxy/src/main/java/com/bytenya/minecraft/velocityguard/limbo/server/commands/CmdConn.java
@@ -0,0 +1,31 @@
+package com.bytenya.minecraft.velocityguard.limbo.server.commands;
+
+import com.bytenya.minecraft.velocityguard.limbo.server.Command;
+import com.bytenya.minecraft.velocityguard.limbo.server.LimboServer;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+public class CmdConn implements Command {
+    private static final Logger logger = LogManager.getLogger(CmdConn.class);
+
+    private final LimboServer server;
+
+    public CmdConn(LimboServer server) {
+        this.server = server;
+    }
+
+    @Override
+    public void execute() {
+        logger.info("Connections: {}", server.getConnections().getCount());
+    }
+
+    @Override
+    public String name() {
+        return "conn";
+    }
+
+    @Override
+    public String description() {
+        return "Display connections count";
+    }
+}
