diff --git a/proxy/src/main/java/com/bytenya/minecraft/velocityguard/limbo/protocol/packets/play/PacketSpawnPosition.java b/proxy/src/main/java/com/bytenya/minecraft/velocityguard/limbo/protocol/packets/play/PacketSpawnPosition.java
new file mode 100644
index 00000000..35fcbff2
--- /dev/null
+++ b/proxy/src/main/java/com/bytenya/minecraft/velocityguard/limbo/protocol/packets/play/PacketSpawnPosition.java
@@ -0,0 +1,30 @@
+package com.bytenya.minecraft.velocityguard.limbo.protocol.packets.play;
+
+import com.bytenya.minecraft.velocityguard.limbo.protocol.ByteMessage;
+import com.bytenya.minecraft.velocityguard.limbo.protocol.PacketOut;
+import com.bytenya.minecraft.velocityguard.limbo.protocol.registry.Version;
+
+public class PacketSpawnPosition implements PacketOut {
+
+    private long x;
+    private long y;
+    private long z;
+
+    public PacketSpawnPosition() { }
+
+    public PacketSpawnPosition(long x, long y, long z) {
+        this.x = x;
+        this.y = y;
+        this.z = z;
+    }
+
+    @Override
+    public void encode(ByteMessage msg, Version version) {
+        msg.writeLong(encodePosition(x, y ,z));
+        msg.writeFloat(0);
+    }
+
+    private static long encodePosition(long x, long y, long z) {
+        return ((x & 0x3FFFFFF) << 38) | ((z & 0x3FFFFFF) << 12) | (y & 0xFFF);
+    }
+}
